pennylane/gradients/hessian_transform.py

```diff
@@ -0,0 +1,223 @@
+
+    .. code-block:: python
+
+        @hessian_transform
+        def my_custom_hessian(tape, **kwargs):
+            ...
+            return hessian_tapes, processing_fn
+
+    where:
+
+    - ``tape`` (*QuantumTape*): the input quantum tape to compute the Hessian of
+
+    - ``hessian_tapes`` (*List[QuantumTape]*): is a list of output tapes to be
+      evaluated. If this list is empty, no quantum evaluations will be made.
+
+    - ``processing_fn`` is a processing function to be applied to the output of the
+      evaluated ``hessian_tapes``. It should accept a list of numeric results with
+      length ``len(hessian_tapes)``, and return the Hessian matrix.
+
+    Once defined, the quantum Hessian transform can be used as follows:
+
+    >>> hessian_tapes, processing_fn = my_custom_hessian(tape, *hessian_kwargs)
+    >>> res = execute(tapes, dev, interface="autograd", gradient_fn=qml.gradients.param_shift)
+    >>> jacobian = processing_fn(res)
+
+    Alternatively, Hessian transforms can be applied directly to QNodes, in which case
+    the execution is implicit:
+
+    >>> fn = my_custom_gradient(qnode, *gradient_kwargs)
+    >>> fn(weights)  # transformed function takes the same arguments as the QNode
+    1.2629730888100839
+
+    .. note::
+
+        The input tape might have parameters of various types, including NumPy arrays,
+        JAX DeviceArrays, and TensorFlow and PyTorch tensors.
+
+        If the Hessian transform is written in a autodiff-compatible manner, either by
+        using a framework such as Autograd or TensorFlow, or by using ``qml.math`` for
+        tensor manipulation, then higher-order derivatives will also be supported.
+
+        Alternatively, you may use the ``tape.unwrap()`` context manager to temporarily
+        convert all tape parameters to NumPy arrays and floats:
+
+        >>> with tape.unwrap():
+        ...     params = tape.get_parameters()  # list of floats
+    """
+
+    def __init__(
+        self, transform_fn, expand_fn=expand_invalid_trainable, differentiable=True, hybrid=True
+    ):
+        self.hybrid = hybrid
+        super().__init__(transform_fn, expand_fn=expand_fn, differentiable=differentiable)
+
+    @staticmethod
+    def _jacobian_trainable_args(args, interface):
+        """Return the indices of QNode arguments for which a Jacobian was
+        computed by `qml.transforms.classical_jacobian` with argnum=None.
+        """
+        trainable_args = []
+
+        if interface == "autograd":
+            for idx, arg in enumerate(args):
+                # TODO: make default False once this change is done in qml.jacobian
+                if getattr(arg, "requires_grad", True):
+                    trainable_args.append(idx)
+
+        elif interface == "jax":
+            trainable_args = [0]
+
+        # Torch and Tensorflow interfaces are not considered since `classical_jacobian`
+        # always returns a tuple for them, thus not invoking this function.
+
```

Comment:
This is a good point, there was actually some discussion relating to this [in the PR that introduced this fix for the gradient transform](https://github.com/PennyLaneAI/pennylane/pull/1989#discussion_r766067962). Originally, I included the behaviour of all four interfaces for completeness/safety:
```py
        trainable_args = []

        if interface == "autograd":
            for idx, arg in enumerate(args):
                if getattr(arg, "requires_grad", True):
                    trainable_args.append(idx)

        elif interface == "torch":
            trainable_args = range(len(args))

        elif interface == "jax":
            trainable_args = [0]

        elif interface == "tf":
            trainable_args = range(len(args))

        return trainable_args
```
, but you run into the problem that you write code that is never supposed to execute in the first place/is not really testable.

I'm a bit cautious regarding assertions because I don't know what our policy is about them in our production software. Excluding tests, I only found 5 instances in the code base that use the assert statement.

Maybe we could raise an error for any unexpected interface:
```py
        else:  # pragma: no cover
            raise ValueError("Unexpected interface in Hessian transform")
```
, or check at the point of use that the trainable argument indices are not empty:
```py
            trainable_args_idx = self._jacobian_trainable_args(args, qnode.interface)
            if not trainable_args_idx:  # pragma: no cover
                warnings.warn(
                    "No trainable indices on the classical Jacobian, could not compute the hybrid "
                    "Hessian of the QNode. You can still attempt to obtain the quantum Hessian "
                    "with the `hybrid=False` parameter.",
                    UserWarning,
                )
                return ()
            # Since all arguments have the same shape, obtain shape from the first trainable arg
            qnode_arg_shape = qml.math.shape(args[trainable_args_idx[0]])
```
The second is similar to the solution we already use [further down](https://github.com/dime10/pennylane/blob/da334b2b1c2cf270174554a97366bb962ad75a1c/pennylane/gradients/hessian_transform.py#L208), which I like because it produces no error but is not silent either.
Thankfully these measures will be obsolete with #2059.

Reply:
Given the very last point you mention (yay :D), this seems a point of minor priority? I agree, assertions are rarely used and I believe discouraged :) I'd suggest to do whatever requires lowest effort here, as it likely willl not appear in any release.